### ISSUE TITLE
closure-compiler cannot find contentEditable on Node

### ISSUE THREAD

**Issue by nicks** [google/closure-compiler#1443]
Code:

``` javascript
/** 
 * @param {Node} node
 * @return {boolean} 
 */
function isContentEditable(node) {
  return node.contentEditable == 'true'
}
```

Expected: no warnings
Actual: 

```
input0:4: WARNING - Property contentEditable never defined on node of type Node
  return node.contentEditable == 'true'
         ^
```

compiled with new type inference. Works fine with old type inference.

**Comment by concavelenz** [comment]
contentEditable is only defined on Element.  The NTI is strict about subclass property references.   This is working as expected.   Generally, there are APIs that help tighten Node to Element (.children vs .childNodes, .firstElementChild, etc).    Note that the IArrayLike changes will actually make this problem worse in both NTI and OTI as more references will be known to be Node (rather than unknown).

**Comment by nicks** [comment]
does anyone use NTI in real projects? This creates thousands of warnings on perfectly reasonable code, including closure-library. Checking `node.contentEditable == 'true'` is semantically correct when the property does not exist.

I understand the "purity" motivation for this, but it seems like it would be hard to justify such a high false-positive rate.

also -- shouldn't this use the `may not be present` version of the warning?

**Comment by nicks** [comment]
Also -- what's the workaround? In a strongly typed language like Java, it would not be semantically correct to use a type-cast to Element, because we don't know that node is an Element. And i don't think `node instanceof Element` will work cross-window.

**Comment by ChadKillingsworth** [comment]
I've experimented with NTI for some of my projects but quickly turn it back off. The sheer number of false-positive warnings (that aren't related to nullability) is daunting.

**Comment by concavelenz** [comment]
There are a couple of options to determine that the node is an element so a cast is appropriate:

goog.dom.isElement
node.nodeType == Node.ELEMENT_NODE 

The NTI will also backoff on a number of "property existence" checks.  NodeUtil.isPropertyTest has the details.

**Comment by dimvar** [comment]
> does anyone use NTI in real projects?

Yes, but only small projects so far. The large projects haven't switched yet.

> This creates thousands of warnings on perfectly reasonable code

Can you give some examples of the warnings that you see, and then maybe we can refine the property check appropriately. Are they mostly for comparisons, like in this example?

For a non-comparison property access, where the property exists only on a subtype, it is appropriate to warn. People frequently complain that the current property checks are too loose.

**Comment by dimvar** [comment]
> The sheer number of false-positive warnings (that aren't related to nullability) is daunting.

@ChadKillingsworth please file bugs for these

**Comment by nicks** [comment]
'Are they mostly for comparisons, like in this example?' - hard to say. A lot of them are follow-ons of comparisons, e.g., if node.tagName == 'IMG', use image-specific properties.

**Comment by ChadKillingsworth** [comment]
@dimvar I just tried it with the latest closure-library release and compiler at head. Things are MUCH improved.

**Comment by dimvar** [comment]
@nicks As you find more cases while using NTI, please collect them and send them our way. I want to improve the property-check without going the too-loose route of the current typechecker.

IIRC, TypeScript and Flow also warn if a property exists only on a subtype, so this is more evidence that this decision can work in practice.
